python/ray/dag/compiled_dag_node.py

```diff
@@ -334,34 +361,52 @@ def _get_or_compile(
             elif isinstance(task.dag_node, InputNode):
                 task.output_channel = Channel(
                     buffer_size_bytes=self._buffer_size_bytes,
-                    num_readers=task.num_readers,
+                    num_readers=task.num_readers if task.num_readers else 1,
                 )
             else:
-                assert isinstance(task.dag_node, MultiOutputNode)
+                assert isinstance(task.dag_node, MultiOutputNode) or isinstance(
+                    task.dag_node, InputAttributeNode
+                )
 
             for idx in task.downstream_node_idxs:
                 queue.append(idx)
 
         for node_idx, task in self.idx_to_task.items():
-            if node_idx == self.input_task_idx:
-                # We don't need to assign an actual task for the input node.
+            if node_idx in self.input_task_idxs:
+                # We don't need to assign an actual task for the input node(s).
                 continue
 
             if node_idx == self.output_task_idx:
-                # We don't need to assign an actual task for the input node.
+                # We don't need to assign an actual task for the output node.
                 continue
 
             resolved_args = []
+            # for channel args, the index of read result to take, -1 stands for
+            # taking the whole input
+            read_idxs = []
+            attr_idx = -1
             has_at_least_one_channel_input = False
             for arg in task.args:
                 if isinstance(arg, DAGNode):
+                    is_ipt_attr = isinstance(arg, InputAttributeNode)
+                    if is_ipt_attr:
+                        arg = arg._dag_input_node
+                        attr_idx += 1
```

Comment:
Hmm I was a bit confused how this works. How do we make sure that the attr_idx here matches up with the order of args passed through the channel?

Reply:
refactored to get rid of the confusing attr_idx, updated to arg idx -1. when there are multiple args, task.args will be ordered in InputNode, InputAttributeNode, InputAttributeNode, ..., given InputNode doesn't carry arg in this case, arg idx - 1 would map to the correct input arg passed through the channel